### PR TITLE
vDPA interface design proposal

### PR DESCRIPTION
**Pull request by makoto126**
* vDPA interface design proposal

### CODE REVIEW

File: design-proposals/vdpa-interfaces.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+metadata:
+  name: offload-ovn
+  namespace: default
+  annotations:
+    k8s.v1.cni.cncf.io/resourceName: intel.com/vdpa_jaguar_vhost
+spec:
+  config: >-
+    {
+        "cniVersion": "0.3.0",
+        "type": "kube-ovn",
```
Comment (zzfancy) [inline design-proposals/vdpa-interfaces.md:93]: typo ?  u wanna use  "sriov" here  ?

Comment (makoto126) [reply design-proposals/vdpa-interfaces.md:93]: @zzfancy it's not a typo. sriov-cni also can be used with [this](https://github.com/redhat-nfvpe/vdpa-deployment#sriov-cni). kube-ovn can offload ovs to our smart nic.

**Comment by aburdenthehand** [comment]
@makoto126 Were you still interested in pursuing this proposal?

**Comment by makoto126** [comment]
@aburdenthehand Yes, I am still interested.

**Comment by AlonaKaplan** [comment]
@makoto126 thanks for the design proposal.
Since kubevirt 1.1 we have an API to introduce new network bindings as plugins.
Please take a look at the [user](https://kubevirt.io/user-guide/virtual_machines/network_binding_plugins/) and [developer](https://github.com/kubevirt/kubevirt/blob/main/docs/network-binding-plugin.md) guides.

Besides some core bindings as Masquerade, Bridge and SRIOV, the rest of the bindings won't be part of kubevirt and will be implemented as plugins.
We already have three network binding plugin examples - Slirp, Macvtap and Passt.

Furthermore, @lmilleri@redhat.com already started working on introducing a vDPA network binding plugin and has a working POC for it.
Leonardo, can you please share your work? Maybe you can collaborate.

**Comment by lmilleri** [comment]
@makoto126
This is my personal repository/branch: https://github.com/lmilleri/kubevirt/tree/vdpa
I have introduced a new provider for running all the vdpa-tailored sriov components in kind and a new vdpa sidecar plugin that follows the kubevirt plugin framework.
Please note I have introduced many dirty hacks to make the PoC working
- custom images for sriov-cni, sriov-network-device-plugin, etc
- hardcoded NIC PF (because not able to run the vdpa tool in a container)
- hardcoded /dev/vhost-vdpa-0 in the vdpa sidecar plugin (because not getting the vdpa env-vars/annotations in the sidecar container, etc)
- etc

Basically, I'm able to create a kubevirt VM that is consuming the /dev/vhost-vdpa-0 device.
No IP configuration yet on the device though.

Please let me know if you need more information.

File: design-proposals/vdpa-interfaces.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+Add a vDPA type to InterfaceBindingMethod:
+```go
+// Represents the method which will be used to connect the interface to the guest.
+// Only one of its members may be specified.
+type InterfaceBindingMethod struct {
+	Bridge     *InterfaceBridge     `json:"bridge,omitempty"`
+	Slirp      *InterfaceSlirp      `json:"slirp,omitempty"`
+	Masquerade *InterfaceMasquerade `json:"masquerade,omitempty"`
+	SRIOV      *InterfaceSRIOV      `json:"sriov,omitempty"`
+	VDPA       *InterfaceVDPA       `json:"vdpa,omitempty"`
```
Comment (fabiand) [inline design-proposals/vdpa-interfaces.md:52]: I do wonder if we can avoid this.

Couldn't we rather work on a "protocl" to let a device plugin "expose" what kind of binding it needs?

Aka internally in kubevirt we enable vDPA, but there is no VM API flag for this. Instead the CNI is signaling (can it set an env var?) to request vDPA. KubeVirt will meet this request if possible.

Comment (fabiand) [reply design-proposals/vdpa-interfaces.md:52]: FWIW - This could then be exteneded to other bindings as well, eventually making it optional for some cnis and bindings.

Comment (fabiand) [reply design-proposals/vdpa-interfaces.md:52]: @AlonaKaplan @EdDev ?

File: design-proposals/vdpa-interfaces.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                    "vdpaType": "vhost"
+                }]
+            }
+        ]
+    }
+```
+
+The following environment variables will be present in the Pod:
+```bash
+# kubectl exec -it virt-launcher-myvmi-zmdbh -- env
```
Comment (fabiand) [inline design-proposals/vdpa-interfaces.md:121]: Following up on my comment above, we would add an env var (example) like

```
KUBEVIRT_RESOURCE_NAME_BINDING_FOR_offload=vdpa
KUBEVIRT_RESOURCE_NAME_BINDING_FOR_offload2=vdpa
```

File: design-proposals/vdpa-interfaces.md (hunk 2 above)
Comment (AlonaKaplan) [reply design-proposals/vdpa-interfaces.md:52]: @fabiand lmillei is working on a vdpa binding plugin. He added a reference to it in [this](https://github.com/kubevirt/community/pull/228#issuecomment-1997241806) comment.

Comment (fabiand) [reply design-proposals/vdpa-interfaces.md:52]: Thanks, I understand: But as this is a proposal for KubeVirt I wnated to check with you as a chair of sig-network what you think about such a new internal API

Comment (AlonaKaplan) [reply design-proposals/vdpa-interfaces.md:52]: I don't see any reason to add it as an internal API for VDPA. The network binding plugin should fit the use case. We are now working on extending the binding plugin API to have downward API so the env vars/device-info will get to the sidecar building the libvirt domain.
1. [Device info](https://docs.google.com/document/d/13fLaqZuM8El_CT5XzxPv-1epKsB1-Spv_jtXveEkkdQ/edit#heading=h.ra95t07p2hox). 
2. vdpa [concrete](https://docs.google.com/document/d/13fLaqZuM8El_CT5XzxPv-1epKsB1-Spv_jtXveEkkdQ/edit#heading=h.vgfhk652zjg) example.

> Couldn't we rather work on a "protocl" to let a device plugin "expose" what kind of binding it needs?
> Aka internally in kubevirt we enable vDPA, but there is no VM API flag for this. Instead the CNI is signaling (can it set an env var?) to request vDPA. KubeVirt will meet this request if possible.

Regarding this suggestion, I"m not sure I understand it.
What do you mean by internally in kubevirt enable vDPA? kuevbirt core code would have to be familiar with vDPA but will have to discover that this is the binding it should use based on the information from the device plugin?
Why do we need this third approach? We already have core bindings like masquerade, bridge and SRIOV and we have the binding plugin mechanism for other bindings.

**Comment by EdDev** [comment]
I think this is now superseded by https://github.com/kubevirt/community/pull/281 .
